client-base/src/main/kotlin/app/cash/backfila/client/spi/BackfilaParameters.kt

```diff
@@ -27,20 +30,35 @@ class BackfilaParametersOperator<T : Any>(
   val parametersClass: KClass<T>
 ) {
   /** Constructor parameters used as defaults when missing to create a new T. */
-  private val constructorParameters: List<KParameter> = parametersClass.primaryConstructor!!.parameters
+  private val constructor: KFunction<T> = fetchConstructor(parametersClass)
 
   fun constructBackfillConfig(
     parameters: MutableMap<String, ByteString>,
     dryRun: Boolean
   ): BackfillConfig<T> {
     val map = mutableMapOf<KParameter, Any>()
-    for (parameter in constructorParameters) {
+    for (parameter in constructor.parameters) {
+      val defaultAnnotation = parameter.findAnnotation<BackfilaDefault>()
+      if (defaultAnnotation != null) {
+        if (parameters.containsKey(defaultAnnotation.name)) {
+          val value = parameters[defaultAnnotation.name]!!
+          map[parameter] = TYPE_CONVERTERS[parameter.type.jvmErasure]!!.invoke(value)
+        } else if (!parameter.isOptional) {
+          val defaultValue = defaultAnnotation.value
+          check(defaultValue != null) { "Parameter data class has a required member $parameter with no provided value." }
```

Comment:
nit: this check does nothing, since `defaultValue` is smart cast as a non-null value by the enclosing condition.